File: LeetCode/Tree/Easy/100. Same Tree/main.cpp

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        
        if (p==nullptr && q==nullptr){
            return true;
        }

        // conditions
        if (p==nullptr || q==nullptr){
            return false;
        }else if ((p->left!=nullptr && q->left==nullptr) || (p->left==nullptr && q->left!=nullptr)){
            return false;
        }else if ((p->right!=nullptr && q->right==nullptr) || (p->right==nullptr && q->right!=nullptr)){
            return false;
        }else if(p->val != q->val){
            return false;
        }

        // traversal
        return isSameTree(p->left,q->left) && isSameTree(p->right,q->right);
    }
};

File: LeetCode/Tree/Easy/100. Same Tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def isSameTree(self,root,subRoot):
        if not root and not subRoot:
            return True
        elif not root or not subRoot:
            return False
        elif root.val!=subRoot.val:
            return False

        return self.isSameTree(root.left,subRoot.left) and self.isSameTree(root.right,subRoot.right)

    def isSubtree(self, root, subRoot):
        
        if not root:
            return False
        
        if self.isSameTree(root,subRoot):
            return True
    
        return self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)
        